### PR TITLE
add vimco example + bernoulli sample layers

### PR DESCRIPTION
**Pull request by skaae**
adds vimco as http://arxiv.org/pdf/1602.06725v1.pdf
lb

prior samples:
prior png

### CODE REVIEW

File: examples/vimco.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+def get_nonlin(nonlin):
+    if nonlin == 'rectify':
+        return lasagne.nonlinearities.rectify
+    elif nonlin == 'very_leaky_rectify':
+        return lasagne.nonlinearities.very_leaky_rectify
+    elif nonlin == 'tanh':
+        return lasagne.nonlinearities.tanh
+    else:
+        raise ValueError('invalid non-linearity \'' + nonlin + '\'')
```
Comment (wuaalb) [inline examples/vimco.py]: Could perhaps be simplified to
```python
def get_nonlin(nonlin):
    return getattr(lasagne.nonlinearities, nonlin)
```

File: examples/vimco.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+batch_norm = args.batch_norm == 'true' or args.batch_norm == 'True'
+res_out = args.outfolder
+nonlin_enc = get_nonlin(args.nonlin_enc)
+nonlin_dec = get_nonlin(args.nonlin_dec)
+nhidden = args.nhidden
+latent_size = args.nlatent
+dataset = args.dataset
+batch_size = args.batch_size
+num_epochs = args.nepochs
+eval_epoch = args.eval_epoch
```
Comment (wuaalb) [inline examples/vimco.py]: `batch_norm` (always on) and `eval_epoch` (tests each epoch) are not used

File: examples/vimco.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+sh_x_train = theano.shared(preprocesses_dataset(train_x), borrow=True)
+sh_x_test = theano.shared(preprocesses_dataset(test_x), borrow=True)
+
+
+### MODEL SETUP
+# Recognition model q(z|x)
+l_in = lasagne.layers.InputLayer((None, num_features))
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_in, num_units=nhidden, name='Q_DENSE1'), name='Q_')
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_enc_h1, num_units=nhidden, name='Q_DENSE2'),name='Q_')
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_enc_h1, num_units=nhidden, name='Q_DENSE3'),name='Q_')
```
Comment (wuaalb) [inline examples/vimco.py]: Is it normal the hidden layers do not have nonlinearities?
(If so, `nonlin_enc`, `nonlin_dec` are not used)

File: examples/vimco.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    L_corr = L.dimshuffle(0, 1, 'x') - get_vimco_baseline(
+        log_pxz - log_qz_given_x)
+    g_lb_inference = T.mean( T.sum(dg(L_corr) * log_qz_given_x) + L)
+    grads_inference = T.grad(-g_lb_inference, q_params)
+
+    grads = grads_model + grads_inference
+
+    a = log_pz + log_px_given_z - log_qz_given_x
+    a_max = T.max(a, axis=2, keepdims=True)
+    LL = T.mean(a_max) + T.mean( T.log( T.mean(T.exp(a-a_max), axis=2) ) )
```
Comment (wuaalb) [inline examples/vimco.py]: Could be simplified a little using `utils.log_mean_exp()`

File: examples/vimco.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+xepochs = []
+for epoch in range(1, 1+num_epochs):
+    start = time.time()
+    np.random.shuffle(train_x)
+    sh_x_train.set_value(preprocesses_dataset(train_x))
+    cost_train_ = train_epoch(lr, eq_samples, iw_samples, batch_size)
+    cost_eval_, output_eval = test_epoch(eq_samples, 1000, batch_size)
+    costs_train += [cost_train_]
+    costs_eval += [cost_eval_]
+    plotsamples(res_out + '/prior.png', sample100_model(
```
Comment (wuaalb) [inline examples/vimco.py]: I'd use `sample.png` rather than `prior.png`; and maybe append epoch number?

**Comment by wuaalb** [comment]
Great, thanks!
I was curious how this model compared to the RBM..

File: examples/vimco.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+# TRAIN LOOP
+# We have made some the code very verbose to make it easier to understand.
+total_time_start = time.time()
+costs_train = []
+costs_eval = []
+LL_test1  = []
+xepochs = []
+for epoch in range(1, 1+num_epochs):
+    start = time.time()
```
Comment (wuaalb) [inline examples/vimco.py]: Not used, also not `total_time_start` above

File: examples/vimco.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+print "Training"
+
+# TRAIN LOOP
+# We have made some the code very verbose to make it easier to understand.
+total_time_start = time.time()
+costs_train = []
+costs_eval = []
+LL_test1  = []
+xepochs = []
```
Comment (wuaalb) [inline examples/vimco.py]: `LL_test1`, `xepochs` not used

**Comment by skaae** [comment]
@wuaalb Thanks! All the unused settings are leftovers from the importance weighted example that I modifed. I'll include your suggestions.

File: examples/vimco.py (hunk 3 above)
Comment (skaae) [inline examples/vimco.py]: They have rectify ? but i forgot to use nonlin_enc/nonlin_dec setting.

File: examples/vimco.py (hunk 5 above)
Comment (skaae) [inline examples/vimco.py]: Yes. I think i removed the sample number because it has to run for several hundred epochs so you'll get a lot of plots :)

File: examples/vimco.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+    z_prior = T.ones_like(z)*np.float32(0.5)
+    log_pz = log_bernoulli(z, z_prior).sum(axis=3)
+    log_px_given_z = log_bernoulli(x, x_mu, eps=epsilon).sum(axis=3)
+
+    # Calculate the LL using log-sum-exp to avoid underflow
+    log_pxz = log_pz + log_px_given_z
+
+    # L is (bs, mc) See definition of L in appendix eq. 14
+    L = log_sum_exp(log_pxz - log_qz_given_x, axis=2) + \
+        T.log(1.0/T.cast(iw_samples, 'float32'))
```
Comment (wuaalb) [inline examples/vimco.py]: I guess this one can also be simplified to `L = log_mean_exp(log_pxz - log_qz_given_x, axis=2)`

**Comment by wuaalb** [comment]
Looks good to me!

I wonder if maybe some of the functions used to compute the loss/gradients (baselines, etc.) are generic enough to extract into the library itself.

**Comment by skaae** [comment]
I think we should leave as example code. I'll test the code once more during the week and merge if nothing unexpected pops up :)

Thanks for the review.

**Comment by casperkaae** [comment]
Great! merged #36
